marine.c

```diff
@@ -638,6 +636,23 @@ gboolean* last_seen(const int* arr, int len) {
     return ret_arr;
 }
 
+unsigned int count_distinct(const int* arr, int len) {
+    unsigned int count = 0;
+    GHashTable *seen_values = g_hash_table_new(g_int_hash, g_int_equal);
+
+    for (int i = len - 1; i >= 0; i--) {
+        if (!g_hash_table_contains(seen_values, (arr + i))) {
+            int *key = g_new(gint, 1);
+            *key = arr[i];
+            g_hash_table_add(seen_values, key);
+            count++;
+        }
+    }
+
+    g_hash_table_destroy(seen_values);
```

Comment:
Is this a memory leak?
According to this https://developer.gnome.org/glib/stable/glib-Hash-Tables.html#g-hash-table-destroy you should free the keys yourself or initialize the hash table with `g_hash_table_new_full`.